#' Import scRNAseq data
#'
#' @export
#'
#' @description This is a function for easily importing multiple scRNAseq runs and storing data in a list after filtering as suggested in Seurat workflow. It takes a dataframe explaining the sample details (see below), finds cell-barcode matrices generated by CellRanger pipeline in a folder, and returns a list containing individual Seurat objects. This is useful for integrating multiple datasets for comparisons
#'
#' @param sample_df A data frame containing ID numbers for individual sequencing runs, sample names, and arbitrary sample-level metadata. This data frame must contain at least two columns named `sample_name` and `run_id`. Other columns must be uniquely named and can contain metadata such as genotype, treatment, sex, etc. The character string under `sample_name` will be appended to cell barcodes in the created Seurat objects to enable downstream integration of multiple datasets.
#'
#' @param data_folder A string to point out the relative path of the folder containing data files. In this folder, filtered cell-barcode matrices for individual runs must be present under separate folders, and the folder names should correspond to the `run_id` column in the `sample_df`.
#'
#' @param min.cells A numeric value for filtering genes that are present less than this number of cells (see Seurat package). It defaults to 3.
#'
#' @param min.features A numeric value for filtering cells that have less than this number of features (see Seurat package). It defaults to 200
#'
#' @param verbose TRUE/FALSE for turning messages on/off regarding the data set that is being processed.
#'
#' @param species character string which can be either "human" or "mouse". This will impact the calculation of the mitochondrial gene percentage. Mouse genes are lower case (ex. mt-...) whereas human genes are upper case (ex. MT-...)
#'
#' @return A list object containing Seurat objects. The list items will be named with data from `run_id`.
#'
#' @examples
#'
#' \dontrun{
#'
#' sample_info <- data.frame(run_id = c("1111X","1112X","1113X"),
#'                           sample_name = c("ctrl", "drug1", "drug2"),
#'                           genotype = c("WT","WT","KO"),
#'                           sex = c("F","F","M"))
#'
#' obj_list <- data_importer(sample_df = sample_info,
#'                           data_folder = "../data/",
#'                           min.cells = 3, min.features = 200,
#'                           verbose = T,
#'                           species ="mouse")
#'
#' }
#'
#'






data_importer <- function(sample_df,
                          data_folder = "",
                          min.cells = 3,
                          min.features = 200,
                          verbose = T,
                          species = "mouse"){


    if(sum(c("run_id", "sample_name") %in% colnames(sample_df)) != 2) stop("Check column names of the sample_df. At least two columns must be present named 'run_id' and 'sample_name'.")

    require(Seurat)

    obj_list <- list()


    for(i in sample_df$run_id){

        start_time <- Sys.time()

        if(verbose) message(paste("Processing sample:", i))

        # Read data
        dat <- Read10X(data.dir = paste0(data_folder, i))

        # Append sample info to cell barcodes
        # This helps making cell barcodes unique across multiple runs by attaching unique sample level information
        colnames(dat) <- paste(sample_df[sample_df$run_id == i,]$sample_name, colnames(dat), sep="_")

        # Create Seurat object
        obj <- CreateSeuratObject(counts=dat,
                                  min.cells = min.cells,
                                  min.features = min.features)

        if(verbose) message("Adding metadata")

        # Keep sample name as metadata
        obj@meta.data <- cbind(obj@meta.data, sample_df[sample_df$run_id == i,colnames(sample_df) !="run_id", drop=F])


        if(species == "mouse") mt_name <- "^mt-" else
            if(species == "human") mt_name <- "^MT-" else
                stop("Please specify 'mouse' or 'human' as species")

        # Add mitochondrial gene frequency
        obj[["percent.mt"]] <- PercentageFeatureSet(obj, pattern = mt_name)

        obj_list[[i]] <- obj

        end_time <- Sys.time()

        if(verbose) message(paste("Time elapsed:", round(difftime(end_time, start_time, units = "secs"), digits = 2)))

    }

    obj_list



}

